### DISCUSSION TITLE
"Surprise completions": How do I prevent automatic, unanticipated and undesired completions?

### DISCUSSION THREAD

**Discussion by bobdobbs** [company-mode/company-mode#1399] [Q&A]
I've installed company-mode, and I'm using it in an emacs webdev environment.
I'm using it with web-mode, scss-mode and js2-mode.

Often company selects a completion for me, unprompted.

I will be starting to type word, and then that word will automatically be expanded into something that I don't want.
No candidate list is shown before selection.

After a random, unanticipated, undesired completion has occured, the cursor goes to the beginning of the completed character string, making the effect even more jarring.

In one instance, in javascript buffers, my cursor will sometimes be resting. The the characters "IO" will appear, and the cursor will bounce behind the newly inserted characters and repeat the process. Sometimes I won't know that this is happening until my js compiler - running in terminal - tells me about, because the Surprise Completions have occured in the space of a second or two when my eyes haven't been on that buffer.

I imagine that company is trying to do something helpful, but it isn't working for me.
This behaviour has made fear of random completions part of my workflow.

I desperately want to deactivate this behaviour.

**Comment by dgutov** [comment]
Do you see anything relevant in the `*Messages*` buffer when this happens? Perhaps some error or an odd "unhandled event" message?

**Comment by bobdobbs** [comment]
If web-mode is active, then when the surprise completions happen something I get the message 'ac-php: unable to resolve project root'.

However, the behaviour also occurs outside of web-mode. And in those cases, no messages are output to the *Messages* buffer when the unexpected completions occur.

**Reply by dgutov** [reply]
> ac-php: unable to resolve project root

This is worth reporting to `ac-php`, I suppose.

> However, the behaviour also occurs outside of web-mode. And in those cases, no messages are output to the Messages buffer when the unexpected completions occur.

Have you tried calling `undo` and repeating your exact actions? Does that lead to an error? It would help if you got a scenario that triggered the problem more or less reliably, even if that's only on your machine.

Does this happen in all major modes (or e.g. only with LSP)? Do you have any third-party frontends for Company enabled, such as `company-posframe` or `company-box`?

If we don't have an error to work off, and you don't see the problem in a "bare" session (where only `company` is loaded, and none of the other customizations; maybe also web-mode), the best recommendation I can give is to "bisect" your configuration: comment off a half of your config, see if the problem occurs, depending on that comment out the other half (or not), and then slice the offending half down and down, until the several lines causing the problem are found.
